clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -803,7 +812,10 @@ static void VisitArrayElements(RangeTy Item, QualType FieldTy,
   int64_t ElemCount = CAT->getSize().getSExtValue();
   std::initializer_list<int>{(handlers.enterArray(), 0)...};
   for (int64_t Count = 0; Count < ElemCount; Count++) {
-    VisitField(nullptr, Item, ET, handlers...);
+    if (ET->isScalarType())
+      VisitScalarField(nullptr, Item, ET, handlers...);
```

Comment:
Why we can't handle it inside `VisitField` function?

Reply:
For the same reason as not being able to visit array elements from VisitField just yet. Once structs are decomposed, everything can be done from VisitField. The updated sources use a macro to make this a bit clearer.